worker/online_restore.go

```diff
@@ -314,13 +321,53 @@ func handleRestoreProposal(ctx context.Context, req *pb.RestoreRequest) error {
 	}
 	glog.Infof("Backup map phase is complete. Map result is: %+v\n", mapRes)
 
-	// Reduce the map to pstore using stream writer.
 	sw := pstore.NewStreamWriter()
-	if err := sw.Prepare(); err != nil {
-		return errors.Wrapf(err, "while preparing DB")
+	// We can not use stream writer for incremental restore, we use badger batch write for it.
+	if req.IncrementalFrom > 0 {
+		// If there is a drop all in between the last restored backup and the incremental backups
+		// then drop everything before restoring incremental backups.
+		if mapRes.shouldDropAll {
+			if err := pstore.DropAll(); err != nil {
+				return errors.Wrap(err, "failed to reduce incremental restore map")
+			}
+		}
+
+		dropAttrs := [][]byte{x.SchemaPrefix(), x.TypePrefix()}
+		for ns := range mapRes.dropNs {
+			prefix := x.DataPrefix(ns)
+			dropAttrs = append(dropAttrs, prefix)
+		}
+		for attr := range mapRes.dropAttr {
+			dropAttrs = append(dropAttrs, x.PredicatePrefix(attr))
+		}
+
+		// Any predicate which is currently in the state but not in the latest manifest should
+		// be dropped. It is possible that the tablet would have been moved in between the last
+		// restored backup and the incremental backups being restored.
+		clusterPreds := schema.State().Predicates()
+		validPreds := make(map[string]struct{})
+		for _, pred := range restorePreds {
+			validPreds[pred] = struct{}{}
+		}
+		for _, pred := range clusterPreds {
+			if _, ok := validPreds[pred]; !ok {
+				dropAttrs = append(dropAttrs, x.PredicatePrefix(pred))
+			}
+		}
+		if err := pstore.DropPrefixBlocking(dropAttrs...); err != nil {
+			return errors.Wrap(err, "failed to reduce incremental restore map")
+		}
+		if err := sw.PrepareIncremental(); err != nil {
+			return errors.Wrapf(err, "while preparing DB")
+		}
+	} else {
```

Comment:
This can go to some local function `prepareForReduce()`. I think that would look cleaner.

if req.Incremental == 0 {
     // prepare
     return
}
// prepare for incremental

Reply:
Done